extensions/positron-reticulate/src/extension.ts

```diff
@@ -123,13 +189,119 @@ export class ReticulateRuntimeManager implements positron.LanguageRuntimeManager
 		}
 	}
 
+	setSessions(hostRSessionId: string, reticulateId: string, session: positron.LanguageRuntimeSession) {
+		let sessionsMap: { reticulateSessionId: string; hostRSessionId: string; reticulateId: string }[] =
+			CONTEXT.workspaceState.get('reticulate-sessions-map', []);
+
+		session.onDidEndSession(() => {
+			// Remove the session from the map when it ends.
+			sessionsMap = sessionsMap.filter((pair) => pair.reticulateSessionId !== session.metadata.sessionId);
+			CONTEXT.workspaceState.update('reticulate-sessions-map', sessionsMap);
+			this._sessions.delete(session.metadata.sessionId);
+		});
+
+		const existingSession = sessionsMap.find((pair) => pair.hostRSessionId === hostRSessionId);
+		if (existingSession) {
+			existingSession.reticulateSessionId = session.metadata.sessionId;
+		} else {
+			sessionsMap.push({
+				reticulateSessionId: session.metadata.sessionId,
+				hostRSessionId: hostRSessionId,
+				reticulateId: reticulateId
+			});
+		}
+
+		this._sessions.set(session.metadata.sessionId, session);
+		CONTEXT.workspaceState.update('reticulate-sessions-map', sessionsMap);
+	}
+
+	getSessions(): Array<{ reticulateSessionId: string; hostRSessionId: string; reticulateId: string }> {
+		const sessionsMap = CONTEXT.workspaceState.get('reticulate-sessions-map', []);
+		return sessionsMap as Array<{ reticulateSessionId: string; hostRSessionId: string; reticulateId: string }>;
+	}
+
 	async restoreSession(runtimeMetadata: positron.LanguageRuntimeMetadata, sessionMetadata: positron.RuntimeSessionMetadata): Promise<positron.LanguageRuntimeSession> {
+		const sessionPromise = new PromiseHandles<positron.LanguageRuntimeSession>();
+		vscode.window.withProgress({
+			location: vscode.ProgressLocation.Notification,
+			title: 'Restoring the Reticulate Python session',
+			cancellable: false
+		}, async (progress, _token) => {
+			let session: ReticulateRuntimeSession | undefined;
+			try {
+				session = await this.restoreSession_(runtimeMetadata, sessionMetadata, progress);
+				sessionPromise.resolve(session as positron.LanguageRuntimeSession);
+			} catch (err) {
+				sessionPromise.reject(err);
+			}
+			if (session) {
+				// Wait for the session to start (or fail to start) before
+				// returning from this callback, so that the progress bar stays up
+				// while we wait.
+				progress.report({ increment: 10, message: 'Waiting to connect' });
+				await session.started.wait();
+			}
```

Comment:
Should this throw/err if no session is returned?

Reply:
That's correct! I moved error handling to this layer, so we can easily see what happens when creating the session fails: https://github.com/posit-dev/positron/pull/7024/commits/dee03a2b66ce179de5539496ead56d507bbe75eb